prov/hook/src/hook.c

```diff
@@ -44,6 +44,53 @@ static char **hooks;
 static size_t hook_cnt;
 
 
+struct hook_fabric *hook_to_fabric(const struct fid *fid)
+{
+	switch (fid->fclass) {
+	case FI_CLASS_FABRIC:
+		return (container_of(fid, struct hook_fabric, fabric.fid));
+	case FI_CLASS_DOMAIN:
+		return (container_of(fid, struct hook_domain, domain.fid)->
+			fabric);
+	case FI_CLASS_AV:
+		return (container_of(fid, struct hook_av, av.fid)->
+			domain->fabric);
+	case FI_CLASS_WAIT:
+		return (container_of(fid, struct hook_wait, wait.fid)->
+			fabric);
+	case FI_CLASS_POLL:
+		return (container_of(fid, struct hook_poll, poll.fid)->
+			domain->fabric);
+	case FI_CLASS_EQ:
+		return (container_of(fid, struct hook_eq, eq.fid)->
+			fabric);
+	case FI_CLASS_CQ:
+		return (container_of(fid, struct hook_cq, cq.fid)->
+			domain->fabric);
+	case FI_CLASS_CNTR:
+		return (container_of(fid, struct hook_cntr, cntr.fid)->
+			domain->fabric);
+	case FI_CLASS_SEP:
+	case FI_CLASS_EP:
+	case FI_CLASS_RX_CTX:
+	case FI_CLASS_SRX_CTX:
+	case FI_CLASS_TX_CTX:
+		return (container_of(fid, struct hook_ep, ep.fid)->
+			domain->fabric);
+	case FI_CLASS_PEP:
+		return (container_of(fid, struct hook_pep, pep.fid)->
+			fabric);
+	case FI_CLASS_STX_CTX:
+		return (container_of(fid, struct hook_stx, stx.fid)->
+			domain->fabric);
+	case FI_CLASS_MR:
+		return (container_of(fid, struct hook_mr, mr.fid)->
+			domain->fabric);
+	default:
+		return NULL;
```

Comment:
probably safe to assert(0) here

Reply:
This was the reason I was checking for NULL fabric in other places. If we force using debug hook only when --enable-debug=yes, then we can use assert(0) here.

Reply:
I don't understand how fabric can be null, regardless of which build/hook is used.  What situation causes this?  Do we still need to allocate 'hooked' objects for all underlying objects, even if we don't hook the calls for it?